myriad-scheduler/src/main/java/org/apache/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java

```diff
@@ -81,83 +83,115 @@ public void onEvent(ResourceOffersEvent event, long sequence, boolean endOfBatch
-        // keep this in case SchedulerState gets out of sync. This should not happen with
-        // synchronizing addNodes method in SchedulerState
-        // but to keep it safe
-        final Set<Protos.TaskID> missingTasks = Sets.newHashSet();
-        Set<Protos.TaskID> pendingTasks = schedulerState.getPendingTaskIds();
-        if (CollectionUtils.isNotEmpty(pendingTasks)) {
-          for (Protos.TaskID pendingTaskId : pendingTasks) {
-            NodeTask taskToLaunch = schedulerState.getTask(pendingTaskId);
-            if (taskToLaunch == null) {
-              missingTasks.add(pendingTaskId);
-              LOGGER.warn("Node task for TaskID: {} does not exist", pendingTaskId);
-              continue;
-            }
-            String taskPrefix = taskToLaunch.getTaskPrefix();
-            ServiceResourceProfile profile = taskToLaunch.getProfile();
-            Constraint constraint = taskToLaunch.getConstraint();
-
-            Set<NodeTask> launchedTasks = new HashSet<>();
-            launchedTasks.addAll(schedulerState.getActiveTasksByType(taskPrefix));
-            launchedTasks.addAll(schedulerState.getStagingTasksByType(taskPrefix));
-
-            ResourceOfferContainer resourceOfferContainer = new ResourceOfferContainer(offer, taskToLaunch.getProfile(), role);
-            if (SchedulerUtils.isUniqueHostname(offer, taskToLaunch, launchedTasks)
-                && resourceOfferContainer.satisfies(taskToLaunch.getProfile(), constraint)) {
-              try {
-                final TaskInfo task = taskFactoryMap.get(taskPrefix).createTask(resourceOfferContainer,
-                    schedulerState.getFrameworkID().get(), pendingTaskId, taskToLaunch);
-                LOGGER.info("Launching task: {} using offer: {}", task.getTaskId().getValue(), offer.getId());
-                LOGGER.debug("Launching task: {} with profile: {} using offer: {}", task, profile, offer);
-                driver.launchTasks(Collections.singleton(offer.getId()), Collections.singleton(task));
-                schedulerState.makeTaskStaging(pendingTaskId);
-                // For every NM Task that we launch, we currently
-                // need to backup the ExecutorInfo for that NM Task in the State Store.
-                // Without this, we will not be able to launch tasks corresponding to yarn
-                // containers. This is specially important in case the RM restarts.
-                taskToLaunch.setExecutorInfo(task.getExecutor());
-                taskToLaunch.setHostname(offer.getHostname());
-                taskToLaunch.setSlaveId(offer.getSlaveId());
-                schedulerState.addTask(pendingTaskId, taskToLaunch);
-                iterator.remove(); // remove the used offer from offers list
-                break;
-              } catch (Throwable t) {
-                LOGGER.error("Exception thrown while trying to create a task for {}", taskPrefix, t);
-              }
-            }
+      dispatchTasks(driver, reserved);
```

Comment:
You could concatenate the list (unreserved add after reserved) and call dispatchTasks just once